content/blog/making-a-slack-bot-with-node-and-mongo/index.md

```diff
@@ -0,0 +1,623 @@
+
+```javascript
+// index.js
+// Initialize using signing secret from environment variables
+const { createEventAdapter } = require('@slack/events-api');
+// Slack requires a secret key to run your bot code. We'll find and figure out this signing secret thing in the next steps
+const slackEvents = createEventAdapter(process.env.SLACK_SIGNING_SECRET);
+const port = process.env.PORT || 3000;
+
+// Attach listeners to events by Slack Event "type". See: https://api.slack.com/events/message.im
+slackEvents.on('message', (event) => {
+  console.log(`Received a message event: user ${event.user} in channel ${event.channel} says ${event.text}`);
+});
+
+// Handle errors (see `errorCodes` export)
+slackEvents.on('error', console.error);
+
+// Start a basic HTTP server
+slackEvents.start(port).then(() => {
+  // Listening on path '/slack/events' by default
+  console.log(`server listening on port ${port}`);
+});
+```
+
+This code is what we'll need to run a `console.log` every time a user sends a message. However, _we'll need to setup more to get this code actually running due to Slack's permissions systems_ and such. For now, we'll save this code to `index.js` in the same folder we saved our `package.json` file.
+
+Another thing that was mentioned in the code sample was the `process.env.SLACK_SIGNING_SECRET`. This is the key that Slack will use to connect your code to your workspace. We'll want to keep in mind how to store the signing secret (as the name implies, _we want to keep this key a secret as otherwise anyone can highjack your Slack app_). As the above code hints at, it's suggested to use an environmental file or configuration.
+
+While environmental variables are typically assigned by system configurations, to make development easier, we'll setup a `.env` file with the expected credentials. Then, to inject the `.env` file contents into our `process`, we'll run our code using [the `env-cmd` package](https://www.npmjs.com/package/env-cmd). We'll start by installing the package:
+
+```
+npm i env-cmd
+```
+
+This package will look for a `.env` file and inject it into your command that follows `env-cmd`. So, for example, you can **make a new file called `.env` and place the following contents in it**:
+
+```
+SLACK_SIGNING_SECRET=<SIGNING_SECRET_FROM_HOMESCREEN>
+```
+
+Then, in your `package.json`, you can **edit your `start` command** to reflect the following:
+
+```json
+{
+  "scripts": {
+    "start": "env-cmd node ./index.js"
+  }
+}
+```
+
+Now, whenever your code shows `process.env.SLACK_SIGNING_SECRET`, it'll represent the value present in your `.env` file
+
+# Development Hosting {#development-environment-setup}
+
+In order to have these events called, we'll need to get a public URL to route to our local development server. In order to do this, we can [use `ngrok`](https://github.com/inconshreveable/ngrok) to host a public URL in our local environment:
+
+```
+npm i -D ngrok
+npx ngrok http 3000
```

Comment:
is the `npm install` necessary here? npx will run the package regardless of whether it's installed.

Reply:
It will, but I'd rather leave the instruction in than leave it out. Npx is great but having to grab from the network every time is irksome and prevents me from adding `ngrok` to `script` files.